DockerControlSTM/Src/main.c

```diff
@@ -1,227 +1,346 @@
+
+/* Private variables ---------------------------------------------------------*/
+UART_HandleTypeDef huart3;
+
+/* USER CODE BEGIN PV */
+
+extern uint8_t packet_header[10];
+extern uint8_t packet_body[4096];
+
+extern enum esp_connection_state connection_state;
+
+extern enum esp_cmd cmd;
+extern uint8_t cmd_received;
+
+uint8_t uart_receive[5];
+const uint16_t uart_size = 5;
+
+uint8_t usb_data[40];
+uint8_t usb_received = 0;
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+static void MX_GPIO_Init(void);
+static void MX_USART3_UART_Init(void);
+/* USER CODE BEGIN PFP */
+
+void start_esp() {
+    HAL_UART_Receive_IT(&huart3, uart_receive, uart_size);
+
+    esp_init(&huart3);
+    esp_passthrough(&huart3);
+
+    HAL_UART_AbortReceive_IT(&huart3);
+    HAL_UART_Receive_IT(&huart3, packet_header, sizeof(packet_header));
+}
+
+void HAL_UART_RxCpltCallback(UART_HandleTypeDef *huart) {
+    HAL_UART_AbortReceive_IT(huart);
+    HAL_GPIO_TogglePin(GPIOD, GPIO_PIN_13);
+    if(huart->Instance == USART3) {
+        switch (connection_state) {
+            case WAIT_HEADER:
+            {
+                const uint16_t body_size = esp_process_header(huart);
+                HAL_UART_Receive_IT(huart, packet_body, body_size);
+            }
+                break;
+            case WAIT_BODY:
+                esp_process_body(huart);
+                HAL_UART_Receive_IT(huart, packet_header, sizeof(packet_header));
+                break;
+            case IDLE:
+                CDC_Transmit_FS(uart_receive, uart_size);
+                HAL_UART_Receive_IT(huart, uart_receive, uart_size);
+                break;
+        }
+    }
+}
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART3_UART_Init();
+  MX_USB_DEVICE_Init();
+  /* USER CODE BEGIN 2 */
+
+  HAL_GPIO_WritePin(GPIOD, GPIO_PIN_12, GPIO_PIN_SET);
+  util_log("DockerControl start");
+
+  start_esp();
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+
+// clion not to alert endless loop
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wmissing-noreturn"
+
+    util_log("endless loop");
+    while (1)
+    {
+        if (cmd_received == 1) {
```

Comment:
Here received commands can be accessed. When `cmd_received` is `1`, they are ready to read